invoice/src/amount.rs

```diff
@@ -259,6 +268,41 @@ impl CoinAmount {
             precision,
         }
     }
+
+    pub fn to_value(self) -> u64 {
+        let pow = 10_u64.pow(self.precision as u32);
+        self.int * pow + self.fract
+    }
+}
+
+impl FromStr for CoinAmount {
+    type Err = CoinAmountParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let (int, fract, precision) = match s.split_once('.') {
+            Some((int, fract)) => (
+                int.parse()
+                    .map_err(|_| CoinAmountParseError::InvalidNumber(int.to_string()))?,
+                fract
+                    .parse()
+                    .map_err(|_| CoinAmountParseError::InvalidNumber(fract.to_string()))?,
+                Precision::try_from(fract.len() as u8)
+                    .map_err(|_| CoinAmountParseError::InvalidNumber(fract.len().to_string()))?,
+            ),
+            None => (
+                s.parse()
+                    .map_err(|_| CoinAmountParseError::InvalidNumber(s.to_string()))?,
+                0,
+                Precision::Indivisible,
+            ),
+        };
+
+        Ok(CoinAmount {
+            int,
+            fract,
+            precision,
```

Comment:
Nope, we can't guess precision from the amount: a user should be able invoice using asset with precision 8 for the amount `10.3` and not `10.30000000`, otherwise we will have too much visual noise. This was the exact reason I didn't have `FromStr` for this type. I need to think about how to address this problem.

I tend to think that we need to rename `precision` into a `min_precision` and then everything starts to make sense.